### PR TITLE
Forbid ~predicates in `across/if_any/if_all(.cols = )`

### LINKED ISSUE #5732: tidyselect + where VS if_any/if_all, and error messages
Number one, extremely happy to have the new features for `filter` and/or `mutate()` with `if_any`/`if_all`!!! This feature was not something I expected but am so excited to have!!!

🥳 

There is one problem I'd like to raise and one question:

**Problem**: The error message for using `if_any()` without a column selector results in some potentially bad recommendations (related to my question as well).

**Question**: What is the "correct" way of `select`-ing columns based on values *within* that column?

---

While I understand the use of `if_any` with `filter()`, I'm also interested in the similar option for `select()`. when cleaning data or working with "wide" data that I'm putting into a table I fairly often want to `select()` specific columns that *contain* a specific value/string/range/etc. IE just knowing it's numeric or text is not sufficient, but rather I'm interested in specifics of the column's contents.

I have a reprex below that shows some of my testing steps in trying to answer this question.

``` r
library(dplyr, warn.conflicts = FALSE)
library(palmerpenguins)
#> Warning: package 'palmerpenguins' was built under R version 4.0.2


# Filter specific columns 

ex1 <- penguins %>% 
  filter(if_any(contains("bill"), ~.x == 39.1| .x == 18.7))

ex1
#> # A tibble: 6 x 8
#>   species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g sex  
#>   <fct>   <fct>           <dbl>         <dbl>            <int>       <int> <fct>
#> 1 Adelie  Torge…           39.1          18.7              181        3750 male 
#> 2 Adelie  Biscoe           37.7          18.7              180        3600 male 
#> 3 Adelie  Dream            39            18.7              185        3650 male 
#> 4 Chinst… Dream            45.4          18.7              188        3525 fema…
#> 5 Chinst… Dream            51.5          18.7              187        3250 male 
#> 6 Chinst… Dream            50.2          18.7              198        3775 fema…
#> # … with 1 more variable: year <int>

# I realize this is a toy example 
# that could be solved by normal filter
ex2 <- penguins %>% 
  filter(bill_length_mm == 39.1| bill_depth_mm == 18.7)

all.equal(ex1, ex2)
#> [1] TRUE

# Skip tidyselect ---------------------------------------------------------

# What happens when you don't indicate tidyselect or colname?
# Throws a useful error but in this case leads down
# a not very useful path
penguins %>% 
  filter(if_any(~.x == 39.1))
#> Error: Problem with `filter()` input `..1`.
#> x Formula shorthand must be wrapped in `where()`.
#> 
#>   # Bad
#>   data %>% select(~.x == 39.1)
#> 
#>   # Good
#>   data %>% select(where(~.x == 39.1))
#> ℹ Input `..1` is `if_any(~.x == 39.1)`.

# So then let's follow the recommendation of:
# data %>% select(where(~.x == 39.1))
# this fails because it's not vectorized I believe?

penguins %>% 
  select(where(~.x == 39.1))
#> Error: `where()` must be used with functions that return `TRUE` or `FALSE`.

# Still fails with everything() indicated
penguins %>% 
  select(where(if_any(everything(), ~.x == 39.1)))
#> Error: `across()` must only be used inside dplyr verbs.

# Use base::any ---------------------------------------------------------

# So for this behavior, I want to select a specific column
# by values present in that column
# this WILL work if you use base::any() without any extra work!!!
# But I wanted to make sure this is a good idea?

penguins %>% 
  select(where(~any(.x == 36.7, na.rm = TRUE)))
#> # A tibble: 344 x 1
#>    bill_length_mm
#>             <dbl>
#>  1           39.1
#>  2           39.5
#>  3           40.3
#>  4           NA  
#>  5           36.7
#>  6           39.3
#>  7           38.9
#>  8           39.2
#>  9           34.1
#> 10           42  
#> # … with 334 more rows
```

<sup>Created on 2021-02-03 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #5732

on the original example, this gives: 

``` r
library(dplyr, warn.conflicts = FALSE)

mtcars %>% 
  filter(if_any(~.x == 39.1))
#> Error: Problem with `filter()` input `..1`.
#> x Misplaced formula shorthand.
#> ℹ You most likely meant to use `.fns = ~.x == 39.1`
#> ℹ ... or perhaps `where(~.x == 39.1)`
#> ℹ Input `..1` is `if_any(~.x == 39.1)`.
```

<sup>Created on 2021-04-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
I guess `tidyselect::eval_select()` was designed with only `select()` in mind, perhaps it could instead be given more information and craft a better message, or abort with a class so that the caller would create the messages ?

**Comment by romainfrancois** [comment]
cc @jthomasmock

File: R/across.R [hunk 1]
```diff
@@ -266,7 +266,21 @@ across_setup_impl <- function(cols, fns, names, .caller_env, mask = peek_mask("a
   # mask layer from the quosure environment (#5460)
   cols <- quo_set_env(cols, data_mask_top(quo_get_env(cols), recursive = FALSE, inherit = FALSE))
 
-  vars <- tidyselect::eval_select(cols, data = mask$across_cols())
+  vars <- withCallingHandlers(
+    tidyselect::eval_select(cols, data = mask$across_cols()),
+    error = function(e) {
+      msg <- conditionMessage(e)
+      if (any(grepl("Formula shorthand", msg))) {
```
Comment (hadley) [inline R/across.R:273]: Can we push this upstream into tidyselect so this isn't quite so hacky?

File: R/across.R [hunk 2]
```diff
@@ -266,7 +266,21 @@ across_setup_impl <- function(cols, fns, names, .caller_env, mask = peek_mask("a
...
+        abort(c(
+          "Misplaced formula shorthand.",
```
Comment (lionel-) [inline R/across.R:275]: I think the error message should be a little more explicit and also mention that the first argument selects a set of column, not a set of rows. And then show the syntax for selecting rows on all columns. I think using argument passing will be confusing to many though, so I would show the binary form `if_any(everything(), ~ .x == "foo")`.

**Review by lionel- — COMMENTED** [review]
My first thought was that it would be cleaner if you did the checking upstream in `if_any()` and `if_all()`. If `.fns` is not supplied you could check for a tilde call in `enquo(.cols)`. On further thought a very common case for named functions (as opposed to tilde call) is `is.na`. So I guess instrumenting the error might be the easiest approach.

However, I think we are still planning to relax the restrictions on `.cols` in `across()` and allow predicates that are not wrapped in `where()` since there is no ambiguity in that case (it does not make sense to use `across()` on a single column). This would also be the case for `if_any()` and `if_all()`. When this happens, the instrumentation of `where()` errors will not work anymore. Instead we will have to instrument "vectorised-predicates" errors. I guess this is not an issue. Tracking in r-lib/tidyselect#233.

On further thought (again), you can still do the checking upstream on `enquo(.cols)` for these three cases:

- `~` calls
- `where()` calls
- Symbols (for the future when unwrapped predicates are allowed)

This should cover everything and no instrumentation is needed. WDYT?

**Comment by romainfrancois** [comment]
Intercepting `.cols=` in `if_any()` and `if_all()` before `.cols` is used in `eval_select()`. 

``` r
library(tidyverse)
library(palmerpenguins)

penguins %>% 
  filter(if_any(~.x == 39.1))
#> Error: Problem with `filter()` input `..1`.
#> x Predicate used in lieu of column selection.
#> ℹ The first argument (`.cols=`) selects a set of columns.
#> ℹ The second argument (`.fns=`) operates on each selected columns to choose rows.
#> ℹ You most likely meant: `if_any(everything(), ~.x == 39.1)`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `if_any(where(~.x == 39.1))`.
#> ℹ Input `..1` is `if_any(~.x == 39.1)`.
```

<sup>Created on 2021-04-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by romainfrancois** [comment]
It is probably less of an issue with `across()` itself, but it might happen if the order or arguments are confused: 

``` r
library(tidyverse)

df <- data.frame(a = 1, b = 2, c = 3)
df %>% 
  summarise(across(~mean(.)))
#> Error: Problem with `summarise()` column `..1`.
#> ℹ `..1 = across(~mean(.))`.
#> x Formula shorthand must be wrapped in `where()`.
#> 
#>   # Bad
#>   data %>% select(~mean(.))
#> 
#>   # Good
#>   data %>% select(where(~mean(.)))
```

<sup>Created on 2021-04-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

perhaps this would better be handled here, just before the call to `eval_select()` https://github.com/tidyverse/dplyr/blob/master/R/across.R#L269

**Comment by romainfrancois** [comment]
Using some catching/rethrowing: 

``` r
library(tidyverse)

df <- data.frame(a = 1, b = 2, c = 3)
df %>% 
  summarise(across(~mean(.)))
#> Error: Problem with `summarise()` column `..1`.
#> ℹ `..1 = across(~mean(.))`.
#> x Predicate used in lieu of column selection.
#> ℹ The first argument (`.cols=`) selects a set of columns.
#> ℹ The second argument (`.fns=`) operates on each selected columns to choose rows.
#> ℹ You most likely meant: `across(everything(), ~mean(.))`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `across(where(~mean(.)))`.

df %>% 
  filter(if_any(~ .== 1))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `if_any(~. == 1)`.
#> x Predicate used in lieu of column selection.
#> ℹ The first argument (`.cols=`) selects a set of columns.
#> ℹ The second argument (`.fns=`) operates on each selected columns to choose rows.
#> ℹ You most likely meant: `if_any(everything(), ~. == 1)`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `if_any(where(~. == 1))`.
```

<sup>Created on 2021-04-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by romainfrancois** [comment]
I wonder though if we could skip the catching rethrowing if `across_setup_impl()` "knew" it was called from `if_any()` somehow ?

File: R/across.R [hunk 3]
```diff
@@ -250,6 +261,16 @@ across_setup <- function(cols, fns, names, key, .caller_env) {
   value
 }
 
+abort_invalid_cols_arg <- function(.cols, fn = "across") {
+  abort(c(
+    "Predicate used in lieu of column selection.",
+    i = "The first argument (`.cols=`) selects a set of columns.",
+    i = "The second argument (`.fns=`) operates on each selected columns to choose rows.",
```
Comment (lionel-) [inline R/across.R:268]:
```suggestion
    i = "The first argument `.cols` selects a set of columns.",
    i = "The second argument `.fns` operates on each selected columns to choose rows.",
```

Comment (lionel-) [reply R/across.R:268]: > operates on each selected columns to choose rows

"choosing rows" is only the case in `if_any()` / `if_all()` right? Maybe:

> operates on each selected columns to choose rows

- `across()`: The second argument `.fns`  is applied to each selected column.
- `if_any/all()`: The second argument `.fns` is applied to each selected column to filter rows.

Comment (romainfrancois) [reply R/across.R]: I guess we can just stop after column. `if_any()` might be used outside of `filter()`

**Comment by romainfrancois** [comment]
This has to wait for `eval_select()` or something similar to be fixed in https://github.com/r-lib/tidyselect/issues/233

**Comment by lionel-** [comment]
> This has to wait for eval_select() or something similar to be fixed in r-lib/tidyselect#233

oh I think we can go ahead with this PR. We'll only need r-lib/tidyselect#233 when/if we lift the `where()` restriction in `across()`.

**Comment by romainfrancois** [comment]
``` r
library(tidyverse)

df <- data.frame(a = 1, b = 2, c = 3)
df %>% 
  summarise(across(~mean(.)))
#> Error: Problem with `summarise()` column `..1`.
#> ℹ `..1 = across(~mean(.))`.
#> x Predicate used in lieu of column selection.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
#> ℹ You most likely meant: `across(everything(), ~mean(.))`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `across(where(~mean(.)))`.

df %>% 
  filter(if_any(~ .== 1))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `if_any(~. == 1)`.
#> x Predicate used in lieu of column selection.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
#> ℹ You most likely meant: `if_any(everything(), ~. == 1)`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `if_any(where(~. == 1))`.
```

<sup>Created on 2021-04-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by romainfrancois** [comment]
Now rebased after @lionel- pull request #5857, handling both the expanded and non expanded versions of `if_any()`/ `if_all()` : 

@hadley @jthomasmock I believe these messages capture the problem described in #5732, but maybe are quite a mouthful ? Any suggestions ?

``` r
library(dplyr, warn.conflicts = FALSE)

df <- data.frame(x = 1:10, y = 1:10)

# expanded case
filter(df, if_any(~ .x > 5))
#> Error: Predicate used in lieu of column selection.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
#> ℹ You most likely meant: `if_any(everything(), ~.x > 5)`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `if_any(where(~.x > 5))`.
filter(df, if_all(~ .x > 5))
#> Error: Predicate used in lieu of column selection.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
#> ℹ You most likely meant: `if_all(everything(), ~.x > 5)`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `if_all(where(~.x > 5))`.

# non expanded case
filter(df, !if_any(~ .x > 5))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `!if_any(~.x > 5)`.
#> x Predicate used in lieu of column selection.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
#> ℹ You most likely meant: `if_any(everything(), ~.x > 5)`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `if_any(where(~.x > 5))`.
filter(df, !if_all(~ .x > 5))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `!if_all(~.x > 5)`.
#> x Predicate used in lieu of column selection.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
#> ℹ You most likely meant: `if_all(everything(), ~.x > 5)`.
#> ℹ If this was truly meant as a column selection, you must wrap with: `if_all(where(~.x > 5))`.
```

<sup>Created on 2021-04-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

File: tests/testthat/test-across.R [hunk 4]
```diff
@@ -399,6 +399,19 @@ test_that("if_any() and if_all() respect filter()-like NA handling", {
   )
 })
 
+test_that("if_any() and if_all() aborts when predicate mistakingly used in .cols= (#5732)", {
+  df <- data.frame(x = 1:10, y = 1:10)
+  expect_snapshot(error = TRUE, {
```
Comment (lionel-) [inline tests/testthat/test-across.R:404]: You can now use:

```r
expect_snapshot({
  # Expanded case
  (expect_error(filter(df, if_any(~ .x > 5))))
  (expect_error(filter(df, if_all(~ .x > 5))))

  # Non expanded case
  (expect_error(filter(df, !if_any(~ .x > 5))))
  (expect_error(filter(df, !if_all(~ .x > 5))))
})
```

This has the advantage that any of these not throwing an error will cause a failure. Whereas `expect_snapshot(error = TRUE)` only fails if none of the expressions throw.

Another advantage is flexibility, you can check for an error class if you'd like to (though not relevant here).

**Review by lionel- — APPROVED** [review]
Looks good!

File: tests/testthat/test-across.R (hunk 4 above)
Comment (romainfrancois) [reply tests/testthat/test-across.R]: Thanks

**Comment by romainfrancois** [comment]
Maybe we can drop the "If this was truly..." line ?

**Comment by romainfrancois** [comment]
Settling with: 

``` r
library(dplyr, warn.conflicts = FALSE)

df <- data.frame(x = 1:10, y = 1:10)

# expanded case
filter(df, if_any(~ .x > 5))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `if_any(~.x > 5)`.
#> x Predicate used in lieu of column selection.
#> ℹ You most likely meant: `if_any(everything(), ~.x > 5)`.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
filter(df, if_all(~ .x > 5))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `if_all(~.x > 5)`.
#> x Predicate used in lieu of column selection.
#> ℹ You most likely meant: `if_all(everything(), ~.x > 5)`.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.

# non expanded case
filter(df, !if_any(~ .x > 5))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `!if_any(~.x > 5)`.
#> x Predicate used in lieu of column selection.
#> ℹ You most likely meant: `if_any(everything(), ~.x > 5)`.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
filter(df, !if_all(~ .x > 5))
#> Error: Problem with `filter()` input `..1`.
#> ℹ Input `..1` is `!if_all(~.x > 5)`.
#> x Predicate used in lieu of column selection.
#> ℹ You most likely meant: `if_all(everything(), ~.x > 5)`.
#> ℹ The first argument `.cols` selects a set of columns.
#> ℹ The second argument `.fns` operates on each selected columns.
```

<sup>Created on 2021-04-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by hadley** [comment]
LGTM!

**Pull request merged**